File: multi-gammon-core/src/main/java/org/akoshterek/backgammon/agent/Agent.java
package org.akoshterek.backgammon.agent;

import org.akoshterek.backgammon.board.Board;
import org.akoshterek.backgammon.board.PositionClass;
import org.akoshterek.backgammon.eval.Evaluator;
import org.akoshterek.backgammon.eval.Reward;
import org.akoshterek.backgammon.move.Move;

import java.nio.file.Path;

/**
 * @author Alex
 *         date 02.08.2015.
 */
public interface Agent {
    String getFullName();
    Path getPath();
    int getPlayedGames();

    void startGame();
    void endGame();
    void doMove(Move move);

    default Reward evaluatePosition(final Board board, PositionClass pc) {
        if(PositionClass.isBearoff(pc) && !supportsBearoff()) {
            pc = PositionClass.CLASS_RACE;
        }

        switch (pc) {
            case CLASS_OVER:
                return evalOver(board);
            case CLASS_RACE:
                return evalRace(board);
            case CLASS_CRASHED:
                return evalCrashed(board);
            case CLASS_CONTACT:
                return evalContact(board);
            case CLASS_BEAROFF1:
                return Evaluator.getInstance().evalBearoff1(board);
            case CLASS_BEAROFF2:
                return Evaluator.getInstance().evalBearoff2(board);
            default:
                throw new RuntimeException("Unknown class. How did we get here?");
        }
    }

    void scoreMoves(Move[] moves, int count);
    Reward scoreMove(Move pm);

    default Reward evalOver(final Board board) {
        return Evaluator.getInstance().evalOver(board);
    }
    default Reward evalRace(final Board board) {
        return evalContact(board);
    }
    default Reward evalCrashed(final Board board) {
        return evalContact(board);
    }

    Reward evalContact(Board board);

    boolean isLearnMode();
    void setLearnMode(boolean learn);
    //fixed means it's unable to learn
    boolean isFixed();
    boolean needsInvertedEval();
    void setNeedsInvertedEval(boolean needsInvertedEval);
    boolean supportsSanityCheck();
    void setSanityCheck(boolean sc);
    boolean supportsBearoff();
    void setCurrentBoard(Board board);

    default void load() {}
    default void save() {}
}

File: multi-gammon-core/src/main/java/org/akoshterek/backgammon/agent/Agent.scala
package org.akoshterek.backgammon.agent

import java.nio.file.Path

import org.akoshterek.backgammon.board.{Board, PositionClass}
import org.akoshterek.backgammon.eval.{Evaluator, Reward}
import org.akoshterek.backgammon.move.Move

trait Agent {
  val fullName: String

  val path: Path

  var isLearnMode: Boolean = false
  val supportsSanityCheck: Boolean = false
  val needsInvertedEval: Boolean = false
  val supportsBearoff: Boolean = false

  private var _playedGames: Int = 0

  def playedGames: Int = _playedGames

  protected var fixed: Boolean = true

  protected var _currentBoard: Board = _
  protected var curPC: PositionClass = PositionClass.CLASS_OVER

  def startGame(): Unit = {}

  def endGame(): Unit = {
    if (isLearnMode) {
      _playedGames += 1
    }
  }

  def doMove(move: Move): Unit = {}

  def evaluatePosition(board: Board, pc: PositionClass): Reward = {
    val effectivePc: PositionClass = if (PositionClass.isBearoff(pc) && !supportsBearoff) PositionClass.CLASS_RACE else pc
    effectivePc match {
      case PositionClass.CLASS_OVER => evalOver(board)
      case PositionClass.CLASS_RACE => evalRace(board)
      case PositionClass.CLASS_CRASHED => evalCrashed(board)
      case PositionClass.CLASS_CONTACT => evalContact(board)
      case PositionClass.CLASS_BEAROFF1 => Evaluator.evalBearoff1(board)
      case PositionClass.CLASS_BEAROFF2 => Evaluator.evalBearoff2(board)
      case _ => throw new RuntimeException("Unknown class. How did we get here?")
    }
  }

  def scoreMoves(moves: Array[Move]): Unit = moves.foreach(m => m.arEvalMove = scoreMove(m))

  def scoreMove(pm: Move): Reward = {
    val board: Board = Board.positionFromKey(pm.auch)
    pm.pc = Evaluator.classifyPosition(board)
    evaluatePositionFull(board, pm.pc)
  }

  private def evaluatePositionFull(anBoard: Board, pc: PositionClass): Reward = {
    val reward = new Reward(evaluatePosition(anBoard, pc))
    applySanityCheck(anBoard, reward, pc)
  }

  private def applySanityCheck(anBoard: Board, reward: Reward, pc: PositionClass): Reward = {
    if (!PositionClass.isExact(pc) && supportsSanityCheck && !isLearnMode) {
      Evaluator.sanityCheck(anBoard, reward)
    } else {
      reward
    }
  }

  def evalOver(board: Board): Reward = {
    Evaluator.evalOver(board)
  }

  def evalRace(board: Board): Reward = {
    evalContact(board)
  }

  def evalCrashed(board: Board): Reward = {
    evalContact(board)
  }

  def evalContact(board: Board): Reward

  def currentBoard: Board = _currentBoard

  def currentBoard_=(board: Board): Unit = {
    _currentBoard = board.clone()
    curPC = Evaluator.classifyPosition(currentBoard)
  }

  def load(): Unit = {
  }

  def save(): Unit = {
  }
}
